csi/src/dev.rs

```diff
@@ -0,0 +1,175 @@
+//! the information contained in a device URI from the entries in udev.
+//! Hence there are separate traits for attach and detach that each
+//! device type must implement.
+//!
+//! Attaching a device is performed as follows:
+//! ```ignore
+//!     let uri = "iscsi://192.168.0.20:3260/iqn.2019-05.io.openebs:nexus-11111111-0000-0000-0000-000000000000/0";
+//!     let device = Device::parse(uri)?;
+//!     if let Some(path) = device.find().await? {
+//!         // device already attached
+//!     } else {
+//!         // attach the device
+//!         device.attach().await?;
+//!         // wait for it to show up in udev and obtain the path
+//!         let path = Device::wait_for_device(device, 10).await?;
+//!     }
+//! ```
+//!
+//! Detaching a device is performed via:
+//! ```ignore
+//!     let uuid = Uuid::parse_str(&volume_id)?;
+//!     if let Some(device) = Device::lookup(&uuid).await? {
+//!         device.detach().await?;
+//!     }
+//! ```
+
+use std::{convert::TryFrom, time::Duration};
+
+use tokio::time::delay_for;
+use udev::Enumerator;
+use url::Url;
+use uuid::Uuid;
+
+mod iscsi;
+mod nbd;
+mod nvmf;
+mod util;
+
+const NVME_NQN_PREFIX: &str = "nqn.2019-05.io.openebs";
+
+pub use crate::error::DeviceError;
+pub type DeviceName = String;
+
+#[tonic::async_trait]
+pub trait Attach: Sync + Send {
+    async fn attach(&self) -> Result<(), DeviceError>;
+    async fn find(&self) -> Result<Option<DeviceName>, DeviceError>;
+}
+
+#[tonic::async_trait]
+pub trait Detach: Sync + Send {
+    async fn detach(&self) -> Result<(), DeviceError>;
+    fn devname(&self) -> DeviceName;
+}
+
+pub struct Device;
+
+impl Device {
+    /// Main dispatch function for parsing URIs in order
+    /// to obtain a device implementing the Attach trait.
+    pub fn parse(uri: &str) -> Result<Box<dyn Attach>, DeviceError> {
+        let url = Url::parse(uri).map_err(|error| error.to_string())?;
+        match url.scheme() {
+            "file" => Ok(Box::new(nbd::Nbd::try_from(&url)?)),
+            "iscsi" => Ok(Box::new(iscsi::IscsiAttach::try_from(&url)?)),
+            "nvmf" => Ok(Box::new(nvmf::NvmfAttach::try_from(&url)?)),
+            "nbd" => Ok(Box::new(nbd::Nbd::try_from(&url)?)),
+            scheme => Err(DeviceError::from(format!(
+                "unsupported device scheme: {}",
+                scheme
+            ))),
+        }
+    }
+
+    /// Lookup an existing device in udev matching the given UUID
+    /// to obtain a device implementing the Detach trait.
+    pub async fn lookup(
+        uuid: &Uuid,
+    ) -> Result<Option<Box<dyn Detach>>, DeviceError> {
+        let iscsi_key: String = format!("nexus-{}", uuid.to_string());
+        let nvmf_key: String = format!("uuid.{}", uuid.to_string());
+
+        let mut enumerator = Enumerator::new()?;
+
+        enumerator.match_subsystem("block")?;
+        enumerator.match_property("DEVTYPE", "disk")?;
+
+        for device in enumerator.scan_devices()? {
```

Comment:
I wonder if reversing the logic would simplify these long chains of "if's"

Reply:
I did give this some thought. Reversing the logic requires either more assignments or evaluating each expression twice.
However I think I can make this a bit more readable withought sacrificing efficency by creating a macro

Reply:
Yeah reversing would be an option but.... there are some other things we can do here already:

(1) we can filter out (match_property) to only enumerate Nexus_CAS_Driver, this would would get rid of 3 levels already
(2) grab immediately what you are looking for (ID_ISCSI_SERIAL) its either Some or None, and if its some by virtue of the match_property of (1) we already know we are good, right?

If we do that, our statement would would reduce to:

```rust
match device_property_value("ID_SCSI_SERIAL") {
  Some(serial) => { /* rest of logic */ }
   _ => continue
  }
}

Reply:
I did try to narrow the search by adding multiple filters via match_property(), but found out (the hard way) that the filters do not narrow the search as I had assumed.  The iterator returns records that match *ANY* filter rather than *ALL* filters simultaneuosly.